server/sonar-web/src/main/js/apps/coding-rules/components/CustomRuleCreationModal.tsx

```diff
@@ -0,0 +1,393 @@
+    event.preventDefault();
+    event.currentTarget.blur();
+    this.props.onClose();
+  };
+
+  prepareRequest = () => {
+    /* eslint-disable camelcase */
+    const { customRule, organization, templateRule } = this.props;
+    const params = Object.keys(this.state.params)
+      .map(key => `${key}=${csvEscape(this.state.params[key])}`)
+      .join(';');
+    const ruleData = {
+      markdown_description: this.state.description,
+      name: this.state.name,
+      organization,
+      params,
+      severity: this.state.severity,
+      status: this.state.status
+    };
+    return customRule
+      ? updateRule({ ...ruleData, key: customRule.key })
+      : createRule({
+          ...ruleData,
+          custom_key: this.state.key,
+          prevent_reactivation: !this.state.reactivating,
+          template_key: templateRule.key,
+          type: this.state.type
+        });
+    /* eslint-enable camelcase */
+  };
+
+  handleFormSubmit = (event: React.SyntheticEvent<HTMLFormElement>) => {
+    event.preventDefault();
+    this.setState({ submitting: true });
+    this.prepareRequest().then(
+      newRuleDetails => {
+        if (this.mounted) {
+          this.setState({ submitting: false });
+          this.props.onDone(newRuleDetails);
+        }
+      },
+      (response: Response) => {
+        if (this.mounted) {
+          this.setState({ reactivating: response.status === 409, submitting: false });
+        }
+      }
+    );
+  };
+
+  handleNameChange = (event: React.SyntheticEvent<HTMLInputElement>) => {
+    const { value: name } = event.currentTarget;
+    this.setState((state: State) => {
+      const change: Partial<State> = { name };
+      if (!state.keyModifiedByUser) {
+        change.key = latinize(name).replace(/[^A-Za-z0-9]/g, '_');
+      }
+      return change;
+    });
+  };
+
+  handleKeyChange = (event: React.SyntheticEvent<HTMLInputElement>) => {
+    this.setState({ key: event.currentTarget.value, keyModifiedByUser: true });
+  };
+
+  handleDescriptionChange = (event: React.SyntheticEvent<HTMLTextAreaElement>) => {
+    this.setState({ description: event.currentTarget.value });
+  };
+
+  handleTypeChange = ({ value }: { value: string }) => {
+    this.setState({ type: value });
+  };
+
+  handleSeverityChange = ({ value }: { value: string }) => {
+    this.setState({ severity: value });
+  };
+
+  handleStatusChange = ({ value }: { value: string }) => {
+    this.setState({ status: value });
+  };
+
+  handleParameterChange = (event: React.SyntheticEvent<HTMLInputElement | HTMLTextAreaElement>) => {
+    const { name, value } = event.currentTarget;
+    this.setState((state: State) => ({ params: { ...state.params, [name]: value } }));
+  };
+
+  renderTypeOption = ({ value }: { value: string }) => {
+    return <TypeHelper type={value} />;
+  };
+
+  renderSeverityOption = ({ value }: { value: string }) => {
+    return <SeverityHelper severity={value} />;
+  };
+
+  render() {
```

Comment:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Refactor this method to reduce its Cognitive Complexity from 16 to the 15 allowed. ([typescript:S3776](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=typescript%3AS3776))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Fstas%2Frules&issues=AWEpEOaPzwlWAzALwukZ&open=AWEpEOaPzwlWAzALwukZ)

[ISSUE_KEY]: AWEpEOaPzwlWAzALwukZ